Fix some bugs in refactored BaseAdaptor and adapt UniProtParser to it

mkszuba opened the pull request:
## Description

1. Fix a few subtle bugs which have crept into BaseAdaptor during its refactoring;
2. Adapt UniProtParser to changes in the behaviour of _BaseAdaptor::upload_xref_object_graphs()_ introduced by the refactoring;
3. Fix some bugs which have been discovered in UniProtParser in the course of creating unit tests;
4. Addressed performance issues found in the new UniProtParser, bringing the run time for Swiss-Prot files (haven't tested this version on TrEMBL yet) to very close of how long the old version took;
5. Fix a minor issue in MIMParser;
6. Enable the strict mode of Text::CSV in DBASSParser;
7. Create POD sections for all public methods of DBASSParser, MIMParser, Mim2GeneParser and UniProtParser;
8. Add unit tests for DBASSParser.

## Use case

As you can see this is a rather messy PR because the problem with BaseAdaptor bugs only became known after all other changes from the aforementioned list became known and with several of those changes appearing in the same parts of the source code it would be a bit of a cherry-picking nightmare, then again changes to BaseAdaptor are the only far-reaching ones here. The main point is, having the refactored BaseAdaptor merged into master has resulted in UniProtParser no longer working - partly because of API changes but more importantly due to bugs introduced by the refactoring.

## Benefits

1. UniProtParser, as well as any other parsers using _BaseAdaptor::upload_xref_object_graphs()_, should now work again;
2. UniProtParser is now **much** faster than it was immediately following the rewrite;
3. All identified output discrepancies between the old and new versions of UniProtParser than are not due to bugs in the former should now be gone.

## Possible Drawbacks

Someone will have to review this mess. I suggest keeping commit descriptions in sight, they should help.

## Testing

_Have you added/modified unit tests to test the changes?_

Partly. Work is ongoing on further test suites.

_If so, do the tests pass/fail?_

All the added tests added so pass.

_Have you run the entire test suite and no regression was detected?_

Yes, no regressions observed.

nerdstrike commented on modules/Bio/EnsEMBL/Xref/Parser/DBASSParser.pm:
```diff
@@ -25,26 +25,50 @@ use strict;
+               exception instead.
+  Exceptions : throws on all processing errors
+  Caller     : ParseSource in the xref pipeline
+  Status     : Stable
```
We don't do status any more

nerdstrike commented on modules/Bio/EnsEMBL/Xref/Parser/DBASSParser.pm:
```diff
@@ -164,27 +183,25 @@ sub run {
 
 sub is_file_header_valid {
-  my ( $header ) = @_;
+  my ( @header ) = @_;
```
Brackets not strictly required here, as list context is ensured by the array, but whatever.

mkszuba replied:
I prefer keeping brackets even under such circumstances so that I do not have to do context switching in my mind while extracting function arguments depending on what their type is.

nerdstrike commented on modules/Bio/EnsEMBL/Xref/Parser/MIMParser.pm:
```diff
@@ -26,36 +26,69 @@ use Readonly;
+               exception instead.
+  Exceptions : throws on all processing errors
+  Caller     : ParseSource in the xref pipeline
+  Status     : Stable
```
As before, status is useless without a policy and knowledge to maintain it

nerdstrike commented on modules/t/dbass.t:
```diff
@@ -0,0 +1,334 @@
+=cut
+
+## no critic 'RequireFilenameMatchesPackage'
+package Bio::EnsEMBL::Xref::Test::Parser::DBASSParser;
```
Personally I think naming a package in a test script is overkill, especially is it was probably prompted by the self-same critic rule that no requires a no critic statement.

mkszuba replied:
The effort of doing it is minimal and at least that way I do not have to wonder if the test harness ensures proper separation of individual test suites or will they happily plaster their supposedly internal variables all over _main_. And no, it's not the same critic rule - there is a separate one for a module not having a package declaration :-)

mkszuba commented:
Turns out the warning all three test suites expected to appear has since been removed so they stopped working once I have rebased the branch against master. Fixed, now mim2gene.t and omim.t should pass.

As for dbass.t, I am puzzled - in addition to the above it has the CSV parser complain about inconsistent number of fields in the test files which I a) cannot reproduce on my end, and b) having just manually counted all the commas in both files can state with certainty is not true.
